### ISSUE TITLE
CSF3 doesn't work with Webpack5

### ISSUE THREAD

**Issue by anilanar** [storybookjs/storybook#15972]
**Describe the bug**
CSF3 + Webpack5 builder don't work at the same time.

**To Reproduce**
https://github.com/anilanar/storybook-webpack5-csf3-repro

```sh
> git clone https://github.com/anilanar/storybook-webpack5-csf3-repro.git
> cd storybook-webpack5-csf3-repro
> yarn
> yarn storybook
```

**System**
```
  System:
    OS: Linux 5.10 NixOS 21.05 (Okapi) 21.05.20210721.63ee5cd (Okapi)
    CPU: (12) x64 AMD Ryzen 5 2600 Six-Core Processor
  Binaries:
    Node: 14.17.2 - /nix/store/vp3y4rm082app6f5ld4h9dq5z00bbapp-nodejs-14.17.2/bin/node
    Yarn: 3.0.1 - /nix/store/m8csnvay8fj9479f4zdqnr29s9dbjdha-yarn-1.22.10/bin/yarn
    npm: 6.14.13 - /nix/store/vp3y4rm082app6f5ld4h9dq5z00bbapp-nodejs-14.17.2/bin/npm
  Browsers:
    Firefox: 90.0
  npmPackages:
    @storybook/addon-actions: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
    @storybook/addon-docs: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
    @storybook/addon-essentials: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
    @storybook/addon-links: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
    @storybook/builder-webpack5: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
    @storybook/manager-webpack5: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
    @storybook/react: ^6.4.0-alpha.33 => 6.4.0-alpha.33 
```

**Additional context**
Simply because `path` polyfill included in the default Webpack5 config doesn't contain `path.join` which is used by CSF3's auto title generator. Polyfilling of built-in node.js modules is the primary difference between Webpack4 and Webpack5 so this bug is somewhat understandable.

**Comment by scott-lc** [comment]
I am having this exact same problem in `6.4.0-rc.6` (albeit with Vite, not Webpack). I traced down the issue but do not know how to proceed.

Like @anilanar , I have a single CSF3 story:

```
const Gizmo = (args) => <button {...args} />;

export default {
  component: Gizmo,
  argTypes: {
    backgroundColor: { control: 'color' }
  }
};

export const Primary = {
  args: {
    primary: true,
    label: 'Gizmo'
  }
};
```

Actually, the test itself is irrelevant.  The key thing to note is that there is no named story `function/StoryFn` exported, just a CSF3 story `object/StoryObj`. 

I've believe I've narrowed the issue down to `@storybook/source-loader` [generate-helpers.js#L97](https://github.com/storybookjs/storybook/blob/282e716e518b332325be97fd691807d2f76fc250/lib/source-loader/src/abstract-syntax-tree/generate-helpers.js#L97) call to [traverse-helpers.js#L95](https://github.com/storybookjs/storybook/blob/282e716e518b332325be97fd691807d2f76fc250/lib/source-loader/src/abstract-syntax-tree/traverse-helpers.js#L95)

I've never parsed an AST, but I do see, starting at [traverse-helpers.js#L113](https://github.com/storybookjs/storybook/blob/282e716e518b332325be97fd691807d2f76fc250/lib/source-loader/src/abstract-syntax-tree/traverse-helpers.js#L113)

```
  const isFunctionVariableExport = ...
  const isFunctionDeclarationExport = ...

  if (isFunctionDeclarationExport || isFunctionVariableExport) {
   // NEVER GETS HERE
  }
```

The end result is that `generateSourceWithDecorators` does not "see" any exports and ends up generating an empty story file.

Interestingly, If I add a CSF2 style functional story (`StoryFn`) export to the file, everything works fine, including the CSF3 `StoryObj`story .  The absence of a function variable/declaration export is what causes (or leads to) the issue.

**Comment by sakito21** [comment]
@shilman Can I work on this?

**Comment by shilman** [comment]
@sakito21 yes,  we would love some help on this! Please check out [how to contribute](https://storybook.js.org/docs/react/contribute/how-to-contribute) in our docs and feel free to ask questions in `#contributing` in [our Discord](https://discord.gg/storybook). @tmeasday and I can review your PR!

**Comment by sakito21** [comment]
@shilman I'm joined Discord! I'll try my best!

**Comment by tmeasday** [comment]
@shilman is this PR the fix : https://github.com/storybookjs/storybook/pull/16913?

**Comment by shilman** [comment]
@tmeasday Yes, fixed in #16913 and released in 6.4.8
